### PR TITLE
MINOR: Add log segment unit tests, If the maximum offset beyond index, appen…

### PR DESCRIPTION
**Pull request by lanshiqin**
Add log segment unit tests, If the maximum offset beyond index, appended to the log section, it throws LogSegmentOffsetOverflowException

*More detailed description of change
The unit tests for the log segment should cover more scenario validation, and the unit tests submitted this time cover LogSegmentOffsetOverflowException that are expected to be thrown when the maximum index is exceeded

### CODE REVIEW

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: In the spirit of keeping this test run fast and deterministic, could we hard code a value here?

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: I would suggest to use `@ParameterizedTest` and also test for:
1. Negative value for largestOffset
2. Value exceeding the range of integer (existing)

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: May I suggest to create a segment with non-zero index so that we can test for https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/storage/internals/log/AbstractIndex.java#L554 condition to be false. The `relativeOffset` should fit within an integer but the current test case only tests for situation where we have 0 as starting index.

**Review by divijvaidya — CHANGES_REQUESTED** [review]
Thank you for your contribution! I left some comments so that we can test multiple cases with this same unit test.

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (Hangleton) [inline core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: nit: `largestTimestamp` could be misleading as it could make believe the timestamp itself is set to the maximum possible value.

Comment (lanshiqin) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: Thanks, I should have changed the variable name to the 'currentTime'

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (lanshiqin) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: Thanks, it really shouldn't be hardcoded directly here, I should use @ParameterizedTest

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (lanshiqin) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: Good idea

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (lanshiqin) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: Thanks, good advice, I've modified it along with this scheme

**Comment by lanshiqin** [comment]
Use @ParameterizedTest more test scenarios are covered.
1. Negative value for largestOffset
2. Value exceeding the range of integer 
3. Create a segment with non-zero index so that we can test for https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/storage/internals/log/AbstractIndex.java#L554 condition to be false.

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:77]: Note that baseOffset can be `Long` but `largestOffset - baseOffset` should be <= `Integer.MaxValue`. This case is missing from our test here.

Could we add the following test cases as well:

baseOffset is a number > Integer.MaxValue, largestOffset is a number > Integer.MaxValue, such that `largestOffset - baseOffset` > Integer.MaxValue. This will throw an error.

**Review by divijvaidya — CHANGES_REQUESTED** [review]
Thank you for making the suggested changes. 

One last suggestion to add a missing case.

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (lanshiqin) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:77]: Thanks, I've added this use case data

**Review by divijvaidya — APPROVED** [review]
Looks good to me! (and welcome to the Kafka community)

The failing tests are unrelated to this test case. As a next step, a committer will look into this and merge it. Perhaps @showuon will be interested.

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (showuon) [inline core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: This comment is not clear enough (ex: what is maximum offset mean in this test? what index mean here?). Maybe:
```
LogSegmentOffsetOverflowException should be thrown while appending the logs if:
1. largestOffset - baseOffset < 0
2. largestOffset - baseOffset > Integer.MAX_VALUE
```

**Review by showuon — COMMENTED** [review]
LGTM, just a minor comment.

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (lanshiqin) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: Good idea. I've modified it

**Review by showuon — APPROVED** [review]
LGTM! Will merge it after CI build completed. Thanks for the PR!

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (Hangleton) [inline core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: Nit: we can remove (1) since (1) => (2).

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (Hangleton) [inline core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:86]: What is the expected behaviour for `baseOffset == largestOffset`?

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (Hangleton) [inline core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:84]: nit: Note that the max int value on the Java platform is `2147483647` so we are exercising max int + 1.

**Review by Hangleton — COMMENTED** [review]
Thanks for the PR. A few minor comments.

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (lanshiqin) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:86]: When baseOffset == largestOffset, the teardown method in the test class is called at the end of the test, which in turn calls the logSegment _.close() => timeIndex.maybeAppend, Eventually an exception will be thrown (Integer overflow for offset: 0), but this exception is not the point that this test method needs to cover

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (lanshiqin) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:84]: Yes, it was intentionally written as 2147483648. The test method and LogSegment receive offset arguments of type Long in order to test scenarios exceeding Int.MAX_VALUE

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (lanshiqin) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala]: Thanks for your advice, I have revised the description.

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (Hangleton) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:86]: I see, thanks for checking. Could we maybe confirm if the overflow exception comes from this line in `TimeIndex#maybeAppend`:

```
mmap.putInt(relativeOffset(offset))
```

And if that is the case, why is the relative offset < 0 or > max_int?

Comment (Hangleton) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:86]: I see, thanks for checking. Could we maybe confirm if the overflow exception comes from this line in `TimeIndex#maybeAppend`:

```
mmap.putInt(relativeOffset(offset))
```

And if that is the case, why is the relative offset < 0 or > max_int?

**Review by Hangleton — COMMENTED** [review]
Thanks @lanshiqin for your update. Added a follow-up comment, minor but I am curious since the case you mention involves an offset overflow as well.

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (lanshiqin) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:86]: Sorry, I need to correct that. 
When baseOffset == largestOffset, LogSegment.append executes normally without throwing an exception. Because my unit test of parameters shallowOffsetOfMaxTimestamp value is 0, which can lead to rear unit test processing, TimeIndex# maybeAppend: In a `relativeOffset(offset),` the value of offset is 0. If baseOffset is a value greater than 0, the result will be <0, resulting in an exception.
I have repaired the shallowOffsetOfMaxTimestamp value of this test.

**Comment by showuon** [comment]
@Hangleton , do you have any other comments? If no, I'm going to merge it this week.

File: core/src/test/scala/unit/kafka/log/LogSegmentTest.scala
Comment (Hangleton) [reply core/src/test/scala/unit/kafka/log/LogSegmentTest.scala:86]: Got it, thanks for confirming!

**Review by Hangleton — APPROVED** [review]

**Comment by Hangleton** [comment]
@showuon Thanks for the follow-up and apologies for the delay, the PR seems good to me.

**Pull request merged**
